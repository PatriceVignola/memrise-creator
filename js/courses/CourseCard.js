/* @flow */

import React from 'react';
import {
  Card,
  CardActions,
  CardContent,
  CardCover,
  Title,
  Button,
} from 'react-native-paper';
import type { Course } from '../reducers/courses';

type Props = {
  course: Course,
};

function CourseCard(props: Props) {
  return (
    <Card>
      <CardContent>
        <Title>{props.course.name}</Title>
      </CardContent>
      <CardCover source={{ uri: props.course.photoUrl }} />
      <CardActions>
        <Button>Cancel</Button>
        <Button>OK</Button>
      </CardActions>
    </Card>
  );
}

export default CourseCard;
