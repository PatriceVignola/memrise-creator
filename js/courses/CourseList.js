/* @flow */

import React from 'react';
import { FlatList } from 'react-native';
import { connect } from 'react-redux';
import { compose, withProps, withHandlers } from 'recompose';
import type { NavigationScreenProp, NavigationRoute } from 'react-navigation';

import CourseCard from './CourseCard';
import { fetchCoursesAsync, fetchCurrentUserAsync } from '../actions/fetchApi';
import { getCourses } from '../store/selectors';
import type { Course, ThunkAction } from '../actions/types';
import type { State } from '../store/selectors';

type HigherOrderProps = {
  courses: Course[],
  navigation: NavigationScreenProp<NavigationRoute>,
};

type Props = {
  courses: Course[],
  fetchCurrentUser: () => ThunkAction,
  fetchCourses: () => ThunkAction,
  editCourse: (course: Course) => void,
};

// Name-export the unconnected component for tests
export class CourseList extends React.Component<Props> {
  constructor(props: Props) {
    super(props);

    props.fetchCurrentUser();
    props.fetchCourses();
  }

  render() {
    return (
      // TODO: Put loading gif while the courses are loading
      <FlatList
        data={this.props.courses}
        renderItem={({ item }) => (
          <CourseCard
            course={item}
            onPress={this.props.editCourse}
          />
        )}
        keyExtractor={item => item.id.toString()}
      />
    );
  }
}

const enhance = compose(
  withHandlers({
    editCourse: (props: HigherOrderProps) => (course: Course) => {
      props.navigation.navigate('CourseEdit', { course });
    },
  }),
  withProps((props: HigherOrderProps) => ({
    courses: props.courses,
  })),
);

export default connect((state: State) => ({
  courses: getCourses(state),
}), {
  fetchCourses: fetchCoursesAsync,
  fetchCurrentUser: fetchCurrentUserAsync,
})(enhance(CourseList));
