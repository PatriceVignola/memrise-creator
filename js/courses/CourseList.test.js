/* @flow */

import React from 'react';
import Renderer from 'react-test-renderer';
import ShallowRenderer from 'react-test-renderer/shallow';

import CourseCard from './CourseCard';
import { CourseList } from './CourseList';

const shallowRenderer = new ShallowRenderer();

describe('CourseList', () => {
  const language = {
    id: 1,
    slug: 'Language Slug',
    name: 'Language',
    photoUrl: 'Language Photo Url',
    parentId: 1,
    index: 1,
    languageCode: 'Language Code',
  };

  const course = {
    description: 'Course Description',
    id: 1,
    name: 'Course',
    numLearners: 1,
    numLevels: 1,
    numWords: 1,
    photoUrl: 'Course Photo Url',
    slug: 'Course Slug',
    source: language,
    target: language,
  };

  const mockEditCourseFunction = jest.fn();

  test('renders courses', () => {
    const shallow = shallowRenderer.render((
      <CourseList
        editCourse={jest.fn()}
        courses={[course]}
        fetchCurrentUser={jest.fn()}
        fetchCourses={jest.fn()}
      />
    ));
    expect(shallow).toMatchSnapshot();
  });

  test('navigates to CourseEditScreen when clicking a course card', () => {
    const deep = Renderer.create((
      <CourseList
        editCourse={mockEditCourseFunction}
        courses={[course]}
        fetchCurrentUser={jest.fn()}
        fetchCourses={jest.fn()}
      />
    ));

    deep.root.findByType(CourseCard).props.onPress(course);
    expect(mockEditCourseFunction).toHaveBeenCalledTimes(1);
    expect(mockEditCourseFunction).toHaveBeenCalledWith(course);
  });
});
