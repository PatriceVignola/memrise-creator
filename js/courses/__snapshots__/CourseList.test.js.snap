// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CourseList renders courses 1`] = `
<FlatList
  data={
    Array [
      Object {
        "description": "Course Description",
        "id": 1,
        "name": "Course",
        "numLearners": 1,
        "numLevels": 1,
        "numWords": 1,
        "photoUrl": "Course Photo Url",
        "slug": "Course Slug",
        "source": Object {
          "id": 1,
          "index": 1,
          "languageCode": "Language Code",
          "name": "Language",
          "parentId": 1,
          "photoUrl": "Language Photo Url",
          "slug": "Language Slug",
        },
        "target": Object {
          "id": 1,
          "index": 1,
          "languageCode": "Language Code",
          "name": "Language",
          "parentId": 1,
          "photoUrl": "Language Photo Url",
          "slug": "Language Slug",
        },
      },
    ]
  }
  disableVirtualization={false}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  numColumns={1}
  onEndReachedThreshold={2}
  renderItem={[Function]}
  scrollEventThrottle={50}
  updateCellsBatchingPeriod={50}
  windowSize={21}
/>
`;
